File: 40.combination-sum-ii.py
#
# @lc app=leetcode id=40 lang=python3
#
# [40] Combination Sum II
#
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        
        candidates.sort()
        answer = []
        combination = []
        
        def search(candidates, start, target):
            # 2. constraint
            if target == 0:
                # 3. goal
                answer.append(combination[:])
                return
            
            if target < 0:
                return
            
            # 1. choice
            for i in range(start, len(candidates)):
                if i > start and candidates[i] == candidates[i-1]:
                    continue
                combination.append(candidates[i])
                search(candidates, i+1, target-candidates[i])
                combination.pop()
        
        search(candidates, 0, target)
        return answer
        

File: c++/40.combination_sum_II.cpp
class Solution {
public:
    
    void recur(vector<vector<int>>& answer, vector<int>& items, vector<int>& candidates, int start, int target){
        
        if(target == 0){
            answer.push_back(items);
            return;
        }
        if(target < 0) return;
        
        for(int i = start; i < candidates.size(); i++){
            if(i > start and candidates[i] == candidates[i-1]) continue;
            items.push_back(candidates[i]);
            recur(answer, items, candidates, i+1, target - candidates[i]);
            items.pop_back();
        }
        
    
    }
    
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        
        sort(candidates.begin(), candidates.end());
        
        vector<vector<int>> answer;
        vector<int> items;
        
        recur(answer, items, candidates, 0, target);
        return answer;
        
        
    }
};